File: ir/whileloop_ir.py
from common.span import SourceSpan
from .ir_node import IRNode
from typing import List


class WhileLoopIR(IRNode):
    def __init__(self, test: IRNode, scope_id: int, body_scope_id: int, body: List[IRNode], orelse: List[IRNode], span: SourceSpan=None):
        super().__init__(span)
        self.test = test
        self.scope_id = scope_id
        self.body_scope_id = body_scope_id
        self.body = body
        self.orelse = orelse
        self.span = span

File: rust/src/ir/nodes/whileloop_ir.rs
use crate::ir::{ExprIR, SourceSpan, StmtIR};

#[derive(Debug, Clone)]
pub struct WhileLoopIR {
    pub test: Box<ExprIR>,
    pub scope_id: i64,
    pub body_scope_id: i64,
    pub body: Vec<StmtIR>,
    pub orelse: Vec<StmtIR>,
    pub span: Option<SourceSpan>,
}

impl WhileLoopIR {
    pub fn new(
        test: ExprIR,
        scope_id: i64,
        body_scope_id: i64,
        body: Vec<StmtIR>,
        orelse: Vec<StmtIR>,
        span: Option<SourceSpan>,
    ) -> Self {
        Self {
            test: Box::new(test),
            scope_id,
            body_scope_id,
            body,
            orelse,
            span,
        }
    }
}
